### PR TITLE
Fix Nexus integration

### PR DESCRIPTION
**Pull request by cachedout**
## What does this PR do?

Fixes an error with JSON de-serialization and refactor all Nexus steps to accept a username/password instead of using Vault internally.

## Why is it important?

We want to be able to control the Vault retrieval from the pipeline in a single location instead of letting each step manage it to simplify things and to avoid errors which were happening with authentication.

Additionally, we fixed an issue that was found when de-serializing JSON responses on-the-fly. 

## Related issues
Refs https://github.com/elastic/observability-robots/issues/584

### CODE REVIEW

File: vars/nexusCloseStagingRepository.groovy [hunk 1]
```diff
@@ -58,14 +51,10 @@ def call(Map args = [:]){
   // poll repo activity for close action
   while (true) {
       try {
-          withEnvMask(vars: [
-            [var: "NEXUS_username", password: username],
-            [var: "NEXUS_password", password: password]    ]){
-                conn = Nexus.createConnection(Nexus.getStagingURL(url), env.NEXUS_username, env.NEXUS_password, "repository/${stagingId}/activity")
-            }
+          conn = Nexus.createConnection(Nexus.getStagingURL(url), username, password, "repository/${stagingId}/activity")
```
Comment (v1v) [inline vars/nexusCloseStagingRepository.groovy:54]: I'd say to mask the password

File: vars/nexusCloseStagingRepository.groovy [hunk 2]
```diff
@@ -119,11 +108,7 @@ def call(Map args = [:]){
               }
           }
           Exception exception = new Exception(msg.join('\n'))
-          withEnvMask(vars: [
-            [var: "NEXUS_username", password: username],
-            [var: "NEXUS_password", password: password]    ]){
-                conn = Nexus.createConnection(url, env.NEXUS_username, env.NEXUS_password, "profiles/${stagingProfileId}/drop")
-            }
+          conn = Nexus.createConnection(url, username, password, "profiles/${stagingProfileId}/drop")
```
Comment (v1v) [inline vars/nexusCloseStagingRepository.groovy:111]: ditto

File: vars/nexusCreateStagingRepository.groovy [hunk 3]
```diff
@@ -51,11 +41,7 @@ def call(Map args = [:]){
   int attemptNumber = 0
 
   while (attemptNumber <= retries) {
-      withEnvMask(vars: [
-        [var: "NEXUS_username", password: username],
-        [var: "NEXUS_password", password: password]    ]){
-            conn = Nexus.createConnection(Nexus.getStagingURL(url), env.NEXUS_username, env.NEXUS_password, "profiles/${stagingProfileId}/start")
-        }
+      conn = Nexus.createConnection(Nexus.getStagingURL(url), username, password, "profiles/${stagingProfileId}/start")
```
Comment (v1v) [inline vars/nexusCreateStagingRepository.groovy:44]: ditto

File: vars/nexusCreateStagingRepository.groovy [hunk 4]
```diff
@@ -77,18 +63,10 @@ def call(Map args = [:]){
   // Reset retry counter
   attemptNumber = 0
   
-  withEnvMask(vars: [
-    [var: "NEXUS_username", password: username],
-    [var: "NEXUS_password", password: password]    ]){
-        conn = Nexus.createConnection(Nexus.getStagingURL(url), env.NEXUS_username, env.NEXUS_password, "repository/${stagingId}")
-    }
+  conn = Nexus.createConnection(Nexus.getStagingURL(url), username, password, "repository/${stagingId}")
```
Comment (v1v) [inline vars/nexusCreateStagingRepository.groovy:66]: ditto

File: vars/nexusDropStagingRepository.groovy [hunk 5]
```diff
@@ -30,31 +29,18 @@ import net.sf.json.JSONArray
...
     final int retries = 20
     int attemptNumber = 0
 
     while (attemptNumber < retries) {
-        withEnvMask(vars: [
-            [var: "NEXUS_username", password: username],
-            [var: "NEXUS_password", password: password]    ]){
-                conn = Nexus.createConnection(Nexus.getStagingURL(url), env.NEXUS_username, env.NEXUS_password, "profiles/${stagingProfileId}/drop")
-            }
+        conn = Nexus.createConnection(Nexus.getStagingURL(url), username, password, "profiles/${stagingProfileId}/drop")
```
Comment (v1v) [inline vars/nexusDropStagingRepository.groovy:43]: ditto

File: vars/nexusFindStagingId.groovy [hunk 6]
```diff
@@ -20,36 +20,25 @@
...
-  def password = data?.password
 
   def HttpURLConnection conn
   String stagingURL = Nexus.getStagingURL(url)
-  withEnvMask(vars: [
-    [var: "NEXUS_username", password: username],
-    [var: "NEXUS_password", password: password]    ]){
-        conn = Nexus.createConnection(stagingURL, env.NEXUS_username, env.NEXUS_password, "profile_repositories/${stagingProfileId}")
-    }
+  conn = Nexus.createConnection(stagingURL, username, password, "profile_repositories/${stagingProfileId}")
```
Comment (v1v) [inline vars/nexusFindStagingId.groovy:41]: ditto

File: vars/nexusReleaseStagingRepository.groovy [hunk 7]
```diff
@@ -52,11 +44,7 @@ def call(Map args = [:]) {
   int attemptNumber = 0
 
   while (attemptNumber < retries) {
-      withEnvMask(vars: [
-        [var: "NEXUS_username", password: username],
-        [var: "NEXUS_password", password: password]    ]){
-            conn = Nexus.createConnection(Nexus.getStagingURL(url), env.NEXUS_username, env.NEXUS_password, "profiles/${stagingProfileId}/promote")
-        }
+      conn = Nexus.createConnection(Nexus.getStagingURL(url), username, password, "profiles/${stagingProfileId}/promote")
```
Comment (v1v) [inline vars/nexusReleaseStagingRepository.groovy:47]: ditto

File: vars/nexusUploadStagingArtifact.groovy [hunk 8]
```diff
@@ -58,11 +50,7 @@ def call(Map args = [:]){
 
   File md5_f = Nexus.generateHashFile(fh, 'md5')
   File sha1_f = Nexus.generateHashFile(fh, 'sha1')
-  withEnvMask(vars: [
-    [var: "NEXUS_username", password: username],
-    [var: "NEXUS_password", password: password]    ]){
-      Nexus.upload(stagingURL, env.NEXUS_username, env.NEXUS_password, path, fh)
-    }
+  Nexus.upload(stagingURL, username, password, path, fh)
```
Comment (v1v) [inline vars/nexusUploadStagingArtifact.groovy:53]: ditto

**Review by v1v — COMMENTED** [review]
Since this is a breaking change, I'd say to keep both implementations (vault or user/pass) for backward compatibilities if possible

**Review by mdelapenya — APPROVED** [review]

**Comment by cachedout** [comment]
Thanks for the detailed review, @v1v .

I agree with your proposed approach. I did a partial revert and then modified my approach to preserve the masking. Please have another look and let me know what you think. Thanks!

File: vars/nexusCreateStagingRepository.groovy [hunk 9]
```diff
@@ -52,10 +53,10 @@ def call(Map args = [:]){
 
   while (attemptNumber <= retries) {
       withEnvMask(vars: [
-        [var: "NEXUS_username", password: username],
-        [var: "NEXUS_password", password: password]    ]){
-            conn = Nexus.createConnection(Nexus.getStagingURL(url), env.NEXUS_username, env.NEXUS_password, "profiles/${stagingProfileId}/start")
-        }
+      [var: "NEXUS_username", password: username],
+      [var: "NEXUS_password", password: password]    ]){
+        conn = Nexus.createConnection(Nexus.getStagingURL(url), env.NEXUS_username, ,env.NEXUS_password, "profiles/${stagingProfileId}/start")
```
Comment (v1v) [inline vars/nexusCreateStagingRepository.groovy:58]:
```suggestion
        conn = Nexus.createConnection(Nexus.getStagingURL(url), env.NEXUS_username, env.NEXUS_password, "profiles/${stagingProfileId}/start")
```

File: vars/nexusCreateStagingRepository.txt [hunk 10]
```diff
@@ -4,14 +4,12 @@ Create a Nexus staging repository
 nexusCreateStagingRepository(
   stagingProfileId: my_profile,
   description: "My new staging repo",
-  secret: "secret/release/nexus",
```
Comment (v1v) [inline vars/nexusCreateStagingRepository.txt:7]: This argument exists

File: vars/nexusFindStagingId.txt [hunk 11]
```diff
@@ -3,15 +3,12 @@ Find a Nexus staging repository
 ```
 nexusFindStagingRepository(
   url: "https://oss.sonatype.org",
-  secret: "secret/release/nexus",
```
Comment (v1v) [inline vars/nexusFindStagingId.txt:6]: ditto

File: vars/nexusReleaseStagingRepository.txt [hunk 12]
```diff
@@ -3,13 +3,11 @@ Release a Nexus staging repository
 ```
 nexusReleaseStagingRepository(
   url: "https://oss.sonatype.org",
-  secret: "secret/release/nexus"
```
Comment (v1v) [inline vars/nexusReleaseStagingRepository.txt:6]: ditto

File: vars/nexusUploadStagingArtifact.txt [hunk 13]
```diff
@@ -3,7 +3,6 @@ Upload an artifact to the Nexus staging repository
 ```
 nexusUploadStagingArtifact(
   url: "https://oss.sonatype.org",
-  secret: "secret/release/nexus",
```
Comment (v1v) [inline vars/nexusUploadStagingArtifact.txt:6]: ditto

File: vars/nexusCloseStagingRepository.txt [hunk 14]
```diff
@@ -3,14 +3,12 @@ Close a Nexus staging repository
 ```
 nexusCreateStagingRepository(
   url: "https://oss.sonatype.org",
-  secret: "secret/release/nexus"
```
Comment (v1v) [inline vars/nexusCloseStagingRepository.txt:6]: ditto

**Review by v1v — APPROVED** [review]
Thanks for being patient with the review :)

**Pull request merged**